### PR TITLE
XTREME: Bucc2018 and Tatoeba

### PR DESCRIPTION
**Pull request by zphang**
Adding Bucc2018 and Tatoeba tasks.

Both tasks require installation of the `faiss` library. I've added some documentation for this under examples, but we can shift this around (are we still keeping the documentation website?). Also both tasks (particularly Bucc) use a good deal of code taken from the XTREME repo. I've marked out the functions that have been adapted from there.

## Results

Below are the scores on the validation set using XLM-R-large, `max_seq_length=256`.

Note: Tatoeba results are lower, and Bucc2018 results are higher than those reported in the XTREME paper, but identical to our previous replication.

### Bucc2018

|de|fr|ru|zh|
|--|--|--|--|
|0.777|0.627|0.792|0.665|

### Tatoeba

af|ar|bg|bn|de|el|es|et|eu|fa|fi|fr|he|hi|hu|id|it|ja|jv|ka|kk|ko|ml|mr|nl|pt|ru|sw|ta|te|th|tl|tr|ur|vi|zh|
|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|
|0.305|0.204|0.390|0.133|0.639|0.189|0.480|0.258|0.199|0.420|0.415|0.481|0.280|0.383|0.425|0.470|0.423|0.418|0.102|0.118|0.174|0.355|0.194|0.152|0.526|0.472|0.421|0.079|0.091|0.197|0.274|0.103|0.378|0.225|0.383|0.412|

### CODE REVIEW

File: jiant/tasks/lib/bucc2018.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    Batch = Batch
+
+    TASK_TYPE = TaskTypes.EMBEDDING
+
+    def __init__(self, name, path_dict, language):
+        super().__init__(name=name, path_dict=path_dict)
+        self.language = language
+
+    def get_train_examples(self):
+        raise NotImplementedError()
```
Comment (jeswan) [inline jiant/tasks/lib/bucc2018.py:101]: I'm being pedantic but is this method just not implemented yet or should it never be implemented? If it is the latter, let's change it to a `RuntimeError()`

File: jiant/tasks/lib/bucc2018.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+def get_unique_lines(text_hashes):
+    unique_text_hashes = []
+    unique_indices = []
+    unique_lookup = {}
+    for text_hash in text_hashes:
+        if text_hash not in unique_lookup:
+            unique_lookup[text_hash] = len(unique_lookup)
+            unique_text_hashes.append(text_hash)
+        unique_indices.append(unique_lookup[text_hash])
+    return unique_indices, unique_text_hashes
```
Comment (jeswan) [inline jiant/tasks/lib/bucc2018.py:415]: Seems needlessly complex (unless I am missing something?). Why not convert `text_hashes` to a set and use the corresponding indices?

File: jiant/tasks/lib/bucc2018.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+
+
+def create_ids_map(inds, guids):
+    ids_map = {}
+    for guid, idx in zip(guids, inds):
+        if idx not in ids_map:
+            ids_map[idx] = []
+        ids_map[idx].append(guid)
+    return ids_map
```
Comment (jeswan) [inline jiant/tasks/lib/bucc2018.py:424]: option to use [`defaultdict`](https://docs.python.org/2/library/collections.html#collections.defaultdict) here

File: jiant/tasks/lib/tatoeba.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    TASK_TYPE = TaskTypes.EMBEDDING
+
+    def __init__(self, name, path_dict, language):
+        super().__init__(name=name, path_dict=path_dict)
+        self.language = language
+        self.lang_bimap = labels_to_bimap(["en", language])
+
+    def get_train_examples(self):
+        raise NotImplementedError()
```
Comment (jeswan) [inline jiant/tasks/lib/tatoeba.py:94]: same as above

File: jiant/tasks/lib/tatoeba.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        eng_examples = self._create_examples(
+            lines=read_file_lines(self.path_dict["eng"]), is_english=True, set_type="val",
+        )
+        other_examples = self._create_examples(
+            lines=read_file_lines(self.path_dict["other"]), is_english=False, set_type="val",
+        )
+        return eng_examples + other_examples
+
+    def get_test_examples(self):
+        raise NotImplementedError()
```
Comment (jeswan) [inline jiant/tasks/lib/tatoeba.py:106]: same as above

File: jiant/tasks/lib/tatoeba.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        examples = []
+        for (i, line) in enumerate(lines):
+            examples.append(
+                Example(guid="%s-%s" % (set_type, i), is_english=is_english, text=line,)
+            )
+        return examples
+
+
+def similarity_search(x, y, dim, normalize=False):
+    import faiss
```
Comment (jeswan) [inline jiant/tasks/lib/tatoeba.py:124]: Motivation for `import` here?

File: jiant/tasks/lib/tatoeba.py (hunk 4 above)
Comment (zphang) [reply jiant/tasks/lib/tatoeba.py:94]: Will update.

File: jiant/tasks/lib/tatoeba.py (hunk 5 above)
Comment (zphang) [reply jiant/tasks/lib/tatoeba.py:106]: Will update.

File: jiant/tasks/lib/tatoeba.py (hunk 6 above)
Comment (zphang) [reply jiant/tasks/lib/tatoeba.py:124]: `faiss` is used in this function.

File: jiant/tasks/lib/bucc2018.py (hunk 1 above)
Comment (zphang) [reply jiant/tasks/lib/bucc2018.py:101]: Will update.

File: jiant/tasks/lib/bucc2018.py (hunk 2 above)
Comment (zphang) [reply jiant/tasks/lib/bucc2018.py]: The important part here is to get `unique_indices`, which runs from 0 to \#unique. If I'm understanding your proposal correctly, I think we would still need to create a dictionary mapping `text_hash->idx`. That said, I don't feel strongly either way and can switch over.

File: jiant/tasks/lib/bucc2018.py (hunk 3 above)
Comment (zphang) [reply jiant/tasks/lib/bucc2018.py:436]: I have a (personal) stylistic preference against returning `defaultdict`s (it hides `KeyError`s downstream), although I guess calling `dict(ids_map)` before returning might be fine.

File: jiant/tasks/lib/tatoeba.py (hunk 6 above)
Comment (jeswan) [reply jiant/tasks/lib/tatoeba.py:124]: I'm guessing you are including it here rather than the top of the file since it is an optional library?

File: jiant/tasks/lib/bucc2018.py (hunk 3 above)
Comment (jeswan) [reply jiant/tasks/lib/bucc2018.py:436]: I don't have a strong preference...I'll leave it to your discretion

File: jiant/tasks/lib/bucc2018.py (hunk 2 above)
Comment (jeswan) [reply jiant/tasks/lib/bucc2018.py]: The docstring helps. Here is an alternative implementation which may be simpler to read: https://stackoverflow.com/a/34313968

File: examples/guides/task_specific.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+## Bucc2018, Tatoeba
+
+The Bucc2018 and Tatoeba tasks are sentence retrieval tasks, and require the `faiss` library to run. `faiss-gpu` is recommended for speed reasons.
+
+We recommend running:
+
+```bash
+conda install faiss-gpu cudatoolkit=10.1 -c pytorch
+```
```
Comment (jeswan) [inline examples/guides/task_specific.md:11]: What's the motivation behind not adding `faiss` to requirements.txt?

File: jiant/tasks/lib/tatoeba.py (hunk 6 above)
Comment (zphang) [reply jiant/tasks/lib/tatoeba.py:124]: Yup

File: examples/guides/task_specific.md (hunk 7 above)
Comment (zphang) [reply examples/guides/task_specific.md:11]: `faiss` needs to be compiled for a specific CUDA version, separate from PyTorch. This has caused me some environment issues in the past (e.g. previously it didn't a 10.1 version, and I needed to install it via `--ignore-deps`.) I think they've added a 10.1 compatible version now (although this https://github.com/facebookresearch/faiss/issues/1295 seems to indicate to the contrary, so I'm not sure how it's working on Prince).

Given that it's only used for two tasks, and it has historically had version compatibility issues, I'm inclined to keep it off the requirements list.

File: jiant/tasks/lib/bucc2018.py (hunk 2 above)
Comment (zphang) [reply jiant/tasks/lib/bucc2018.py]: My version looks like the first answer on that page :p In that case, let's just move ahead with the current implementation?
